File: quorum/economics.py
"""Per-run economics: timing + cost for both agents, computed at run time.

Reads the gauntlet-agent's result.json and the coding-agent's frozen
coding-agent-token-usage.json, applies pricing (gauntlet only — the
coding-agent cost is already frozen), and returns a JSON-shaped dict that
the runner persists into verdict.json. Renderers display it verbatim;
they never recompute. PRI-1872.
"""
from __future__ import annotations

import json
from pathlib import Path

from quorum.token_usage import (
    PRICING_ASOF,
    capture_tokens,
    estimate_cost_with,
    pricing_for_model,
)


def _read_gauntlet_result(run_dir: Path) -> dict | None:
    base = run_dir / "gauntlet-agent" / "results"
    if not base.is_dir():
        return None
    for d in base.iterdir():
        rj = d / "result.json"
        if rj.is_file():
            try:
                return json.loads(rj.read_text())
            except (json.JSONDecodeError, OSError):
                return None
    return None


def _gauntlet_block(result: dict) -> dict:
    usage = result.get("usage") or {}
    tokens = {
        "input": int(usage.get("inputTokens", 0) or 0),
        "output": int(usage.get("outputTokens", 0) or 0),
        "cache_create": int(usage.get("cacheCreationInputTokens", 0) or 0),
        "cache_read": int(usage.get("cacheReadInputTokens", 0) or 0),
    }
    tokens["total"] = sum(tokens.values())
    model = (result.get("config") or {}).get("model")
    pricing = pricing_for_model(model)
    est = None
    if pricing is not None:
        est = round(estimate_cost_with({
            "total_input": tokens["input"],
            "total_output": tokens["output"],
            "total_cache_create": tokens["cache_create"],
            "total_cache_read": tokens["cache_read"],
        }, pricing), 6)
    dur = result.get("duration_ms")
    return {
        "duration_ms": int(dur) if isinstance(dur, (int, float)) else None,
        "model": model,
        "tokens": tokens,
        "est_cost_usd": est,
    }


def _coding_block(usage: dict) -> dict:
    # Per-model breakdown (PRI-1872): a coding run is multi-model (main agent
    # + subagents on different models). `models` carries each model's tokens
    # and cost; `est_cost_usd` is their sum (already frozen in the usage file).
    models = []
    for model_id, mt in (usage.get("models") or {}).items():
        models.append({
            "model": model_id,
            "tokens": {
                "input": mt.get("total_input", 0),
                "output": mt.get("total_output", 0),
                "cache_create": mt.get("total_cache_create", 0),
                "cache_read": mt.get("total_cache_read", 0),
                "total": mt.get("total_tokens", 0),
            },
            "est_cost_usd": mt.get("est_cost_usd"),
        })
    models.sort(key=lambda m: m["est_cost_usd"] or 0, reverse=True)
    has_unpriced_model = bool(usage.get("has_unpriced_model")) or any(
        model["est_cost_usd"] is None for model in models
    )
    return {
        "duration_ms": usage.get("duration_ms"),
        "model": usage.get("model"),
        "models": models,
        "tokens": {
            "input": usage.get("total_input", 0),
            "output": usage.get("total_output", 0),
            "cache_create": usage.get("total_cache_create", 0),
            "cache_read": usage.get("total_cache_read", 0),
            "total": usage.get("total_tokens", 0),
        },
        "est_cost_usd": usage.get("est_cost_usd"),
        "has_unpriced_model": has_unpriced_model,
    }


def build_run_economics(run_dir: Path) -> dict | None:
    """Build the economics block for verdict.json, or None if no source exists."""
    result = _read_gauntlet_result(run_dir)
    cu_path = run_dir / "coding-agent-token-usage.json"
    coding_usage = None
    if cu_path.is_file():
        try:
            coding_usage = json.loads(cu_path.read_text())
        except (json.JSONDecodeError, OSError):
            coding_usage = None

    if result is None and coding_usage is None:
        return None

    gauntlet = _gauntlet_block(result) if result is not None else None
    coding = _coding_block(coding_usage) if coding_usage is not None else None

    g_cost = gauntlet["est_cost_usd"] if gauntlet else None
    c_cost = coding["est_cost_usd"] if coding else None
    coding_has_unpriced = bool(coding and coding.get("has_unpriced_model"))
    total = (
        round(g_cost + c_cost, 6)
        if (g_cost is not None and c_cost is not None and not coding_has_unpriced)
        else None
    )

    partial = (gauntlet is None or coding is None
               or g_cost is None or c_cost is None or coding_has_unpriced)

    return {
        "pricing_asof": PRICING_ASOF,
        "gauntlet": gauntlet,
        "coding_agent": coding,
        "total_est_cost_usd": total,
        "partial": partial,
    }


def _detect_backend_and_logs(run_dir: Path) -> tuple[str | None, list[Path]]:
    """Detect coding-agent backend + its session-log files from a run dir.

    The per-run isolated agent-config dir tells us which backend ran:
    Claude writes under `coding-agent-config/projects/`, Codex under
    `coding-agent-config/sessions/`. All logs there belong to this run
    (isolated config), so no cross-run cwd filtering is needed.
    """
    cfg = run_dir / "coding-agent-config"
    projects = cfg / "projects"
    sessions = cfg / "sessions"
    if projects.is_dir():
        return "claude", sorted(projects.rglob("*.jsonl"))
    if sessions.is_dir():
        return "codex", sorted(sessions.rglob("rollout-*.jsonl"))
    return None, []


def backfill_run_economics(run_dir: Path) -> str:
    """Recompute economics for an existing run dir and inject it into verdict.json.

    Regenerates `coding-agent-token-usage.json` from the preserved session
    logs (so it carries the per-model breakdown + corrected cost), then adds
    an `economics` block to the existing `verdict.json`. Returns a short
    status string. NOTE: this re-prices at *current* pricing tables — it is a
    deliberate re-pricing, not a faithful replay of run-time cost.
    """
    backend, logs = _detect_backend_and_logs(run_dir)
    if backend and logs:
        usage = capture_tokens(backend_family=backend, session_log_files=logs)
        if usage is not None:
            (run_dir / "coding-agent-token-usage.json").write_text(
                json.dumps(usage, indent=2) + "\n"
            )

    econ = build_run_economics(run_dir)

    vj = run_dir / "verdict.json"
    if not vj.is_file():
        return "skipped (no verdict.json)"
    try:
        verdict = json.loads(vj.read_text())
    except (json.JSONDecodeError, OSError):
        return "skipped (unreadable verdict.json)"
    verdict["economics"] = econ
    vj.write_text(json.dumps(verdict, indent=2))
    if econ is None:
        return "updated (no economics sources)"
    return "backfilled"

File: ts/src/quorum/economics.ts
/**
 * Per-run economics: timing + cost for both agents, computed at run time.
 *
 * Port of quorum/economics.py — preserves the output dict shape exactly
 * (it goes into verdict.json; renderers display it verbatim and never
 * recompute; the show economics pane and its tests depend on the shape).
 *
 * Reads the gauntlet-agent's usage.jsonl sidecar (priced via obol) and the
 * coding-agent's frozen coding-agent-token-usage.json (already obol-priced
 * at capture time), composes them into the economics block.
 *
 * Every read is best-effort: missing files/fields degrade to null +
 * `partial: true`, never a silent $0.
 */

import * as fs from "fs";
import * as path from "path";
import { estimateUsageSidecar } from "./obol-capture.ts";

/**
 * Find the first gauntlet results subdir that has artifacts.
 * Port of _gauntlet_results_dir.
 */
function gauntletResultsDir(runDir: string): string | null {
  const base = path.join(runDir, "gauntlet-agent", "results");
  if (!fs.existsSync(base) || !fs.statSync(base).isDirectory()) {
    return null;
  }
  const entries = fs
    .readdirSync(base)
    .sort()
    .map((name) => path.join(base, name))
    .filter((d) => fs.statSync(d).isDirectory());

  for (const d of entries) {
    if (
      fs.existsSync(path.join(d, "result.json")) ||
      fs.existsSync(path.join(d, "usage.jsonl"))
    ) {
      return d;
    }
  }
  return null;
}

/**
 * Parse a JSON file into a dict, or null on any error / wrong type.
 * Port of _read_json.
 */
function readJson(filePath: string): Record<string, unknown> | null {
  try {
    const data: unknown = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    if (typeof data === "object" && data !== null && !Array.isArray(data)) {
      return data as Record<string, unknown>;
    }
    return null;
  } catch {
    return null;
  }
}

/**
 * Extract the obol provenance block, or null for pre-obol frozen files.
 * Port of _obol_provenance.
 */
function obolProvenance(
  usage: Record<string, unknown>
): Record<string, unknown> | null {
  if (!("pricing_as_of" in usage)) {
    return null;
  }
  return {
    per_model: (usage["models"] as Record<string, unknown>) ?? {},
    unpriced_models: (usage["unpriced_models"] as string[]) ?? [],
    approximations: (usage["approximations"] as unknown[]) ?? [],
    pricing_as_of: usage["pricing_as_of"],
  };
}

/**
 * Extract aggregate token counts into the shared shell shape.
 * Port of _tokens_shell.
 */
function tokensShell(usage: Record<string, unknown>): Record<string, number> {
  return {
    input: (usage["total_input"] as number) ?? 0,
    output: (usage["total_output"] as number) ?? 0,
    cache_create: (usage["total_cache_create"] as number) ?? 0,
    cache_read: (usage["total_cache_read"] as number) ?? 0,
    total: (usage["total_tokens"] as number) ?? 0,
  };
}

/**
 * Build the gauntlet-agent sub-block.
 * Port of _gauntlet_block.
 */
function gauntletBlock(
  result: Record<string, unknown> | null,
  usage: Record<string, unknown> | null
): Record<string, unknown> | null {
  if (result === null && usage === null) {
    return null;
  }
  const r = result ?? {};
  const u = usage ?? {};
  const dur = r["duration_ms"];
  const config = r["config"];
  const configModel =
    typeof config === "object" && config !== null && !Array.isArray(config)
      ? (config as Record<string, unknown>)["model"]
      : null;
  const model = u["model"] ?? configModel ?? null;
  const hasUnpriced = Boolean(
    Array.isArray(u["unpriced_models"]) &&
      (u["unpriced_models"] as unknown[]).length > 0
  );

  return {
    duration_ms:
      typeof dur === "number" ? Math.trunc(dur) : null,
    model: model ?? null,
    tokens: tokensShell(u),
    est_cost_usd: u["est_cost_usd"] ?? null,
    has_unpriced_model: hasUnpriced,
    obol: usage !== null ? obolProvenance(usage) : null,
  };
}

/**
 * Build the coding-agent sub-block from the frozen token-usage artifact.
 * Port of _coding_block.
 */
function codingBlock(usage: Record<string, unknown>): Record<string, unknown> {
  const rawModels = (usage["models"] as Record<string, Record<string, unknown>> | undefined) ?? {};
  const models: Record<string, unknown>[] = [];

  for (const [modelId, mt] of Object.entries(rawModels)) {
    models.push({
      model: modelId,
      tokens: {
        input: (mt["total_input"] as number) ?? 0,
        output: (mt["total_output"] as number) ?? 0,
        cache_create: (mt["total_cache_create"] as number) ?? 0,
        cache_read: (mt["total_cache_read"] as number) ?? 0,
        total: (mt["total_tokens"] as number) ?? 0,
      },
      est_cost_usd: mt["est_cost_usd"] ?? null,
    });
  }

  // Sort by cost descending; null costs sort last (treated as 0 for the key).
  models.sort((a, b) => {
    const ac = (a["est_cost_usd"] as number | null) ?? 0;
    const bc = (b["est_cost_usd"] as number | null) ?? 0;
    return bc - ac;
  });

  const hasUnpricedFromList = Boolean(
    Array.isArray(usage["unpriced_models"]) &&
      (usage["unpriced_models"] as unknown[]).length > 0
  );
  const hasUnpricedFromModels = models.some(
    (m) => (m["est_cost_usd"] as number | null) === null
  );
  const hasUnpriced = hasUnpricedFromList || hasUnpricedFromModels;

  return {
    duration_ms: usage["duration_ms"] ?? null,
    model: usage["model"] ?? null,
    models,
    tokens: tokensShell(usage),
    est_cost_usd: usage["est_cost_usd"] ?? null,
    tool_result_total_bytes: (usage["tool_result_total_bytes"] as number) ?? 0,
    has_unpriced_model: hasUnpriced,
    obol: obolProvenance(usage),
  };
}

/**
 * Build the economics block for verdict.json, or null if no source exists.
 * Port of build_run_economics.
 */
export async function buildRunEconomics(
  runDir: string
): Promise<Record<string, unknown> | null> {
  const resultsDir = gauntletResultsDir(runDir);
  const gResult = resultsDir
    ? readJson(path.join(resultsDir, "result.json"))
    : null;
  const gUsage = resultsDir
    ? await estimateUsageSidecar(path.join(resultsDir, "usage.jsonl"))
    : null;
  const codingUsage = readJson(
    path.join(runDir, "coding-agent-token-usage.json")
  );

  if (gResult === null && gUsage === null && codingUsage === null) {
    return null;
  }

  const gauntlet = gauntletBlock(gResult, gUsage);
  const coding = codingUsage !== null ? codingBlock(codingUsage) : null;

  const gCost =
    gauntlet !== null ? (gauntlet["est_cost_usd"] as number | null) : null;
  const cCost =
    coding !== null ? (coding["est_cost_usd"] as number | null) : null;
  const codingHasUnpriced = Boolean(
    coding !== null && coding["has_unpriced_model"]
  );
  const gauntletHasUnpriced = Boolean(
    gauntlet !== null && gauntlet["has_unpriced_model"]
  );
  const anyUnpriced = codingHasUnpriced || gauntletHasUnpriced;

  const total =
    gCost !== null && cCost !== null && !anyUnpriced
      ? Math.round((gCost + cCost) * 1_000_000) / 1_000_000
      : null;

  const partial =
    gauntlet === null ||
    coding === null ||
    gCost === null ||
    cCost === null ||
    anyUnpriced;

  let pricingAsof: unknown = null;
  for (const block of [coding, gauntlet]) {
    if (block === null) continue;
    const prov = (block["obol"] as Record<string, unknown> | null) ?? {};
    if (prov["pricing_as_of"]) {
      pricingAsof = prov["pricing_as_of"];
      break;
    }
  }

  return {
    pricing_asof: pricingAsof,
    gauntlet,
    coding_agent: coding,
    total_est_cost_usd: total,
    partial,
  };
}
